validphys2/src/validphys/filters.py

```diff
@@ -168,24 +171,88 @@ def _filter_real_data(filter_path, data):
     return total_data_points, total_cut_data_points
 
 
-def _filter_closure_data(filter_path, data, fakepdf, fakenoise, errorsize):
-    """Filter closure test data."""
+def _filter_closure_data(filter_path, data, fakepdf, fakenoise, filterseed, errorsize):
+    """
+    GENERAL DESCRIPTION:
+
+    This function is accessed within a closure test only, that is, the fakedata
+    namespace has to be True (If fakedata = False, the _filter_real_data function
+    will be used to write the commondata files).
+
+    The function writes commondata and systypes files within the 
+    name_closure_test/filter folder.
+    If fakenoise is True, Level 1 type data is written to the filter folder, otherwise
+    Level 0 data is written.
+    
+    Level 1 data is generated from the Level 0 data by adding noise sampled from 
+    the experimental covariance matrix using the validphys.pseudodata.make_replica
+    function.
+    
+    Parameters
+    ----------
+    
+    filter_path : str
+                  path to filter folder
+
+    data : validphys.core.DataGroupSpec
+
+    fakepdf : validphys.core.PDF 
+
+    fakenoise : bool
+                if fakenoise perform level1 shift of central data values
+
+    filterseed : int
+                 random seed used for the generation of 
+                 random noise added to Level 0 data
+
+    errorsize : float 
+                (defined in runcard)
+    
+
+    Returns
+    -------
+    tuple
+         total data points and points passing the cuts
+
+    """
+    
     total_data_points = 0
     total_cut_data_points = 0
     fakeset = fakepdf.legacy_load()
     # Load data, don't cache result
     loaded_data = data.load.__wrapped__(data)
-    # generate level 1 shift if fakenoise
-    loaded_data.MakeClosure(fakeset, fakenoise)
+
+    from validphys.pseudodata import make_level0_data
+    level0_commondata_instances_wc = make_level0_data(data,fakepdf)
+    commondata_instances_wc = data.load_commondata_instance() # used to generate experimental covariance matrix 
+
     for j, dataset in enumerate(data.datasets):
+        #== print number of points passing cuts, make dataset directory and write FKMASK  ==#
         path = filter_path / dataset.name
         nfull, ncut = _write_ds_cut_data(path, dataset)
         total_data_points += nfull
         total_cut_data_points += ncut
+        # Rescale errors 
         loaded_ds = loaded_data.GetSet(j)
         if errorsize != 1.0:
             loaded_ds.RescaleErrors(errorsize)
-        loaded_ds.Export(str(path))
+    
+    from validphys.commondataparser import write_commondata, write_systype
```

Comment:
Does this create a circular import?

Reply:
It rather prevents it by being inside a function.

Reply:
It does indeed create a circular import error if imported in the header of the module